src/rust/engine/engine_pyo3/src/externs/interface/fs.rs

```diff
@@ -74,6 +86,135 @@ fn match_path_globs(
+
+#[pyclass]
+#[derive(Clone)]
+struct PyDigest(Digest);
+
+#[pymethods]
+impl PyDigest {
+  #[new]
+  fn __new__(fingerprint: &str, serialized_bytes_length: usize) -> PyResult<Self> {
+    let fingerprint = Fingerprint::from_hex_string(fingerprint)
+      .map_err(|e| PyValueError::new_err(format!("Invalid digest hex: {}", e)))?;
+    Ok(Self(Digest::new(fingerprint, serialized_bytes_length)))
+  }
+
+  #[getter]
+  fn fingerprint(&self) -> String {
+    self.0.hash.to_hex()
+  }
+
+  #[getter]
+  fn serialized_bytes_length(&self) -> usize {
+    self.0.size_bytes
+  }
+}
+
+#[pyproto]
+impl PyObjectProtocol for PyDigest {
+  fn __richcmp__(&self, other: PyRef<PyDigest>, op: CompareOp) -> Py<PyAny> {
+    let py = other.py();
+    match op {
+      CompareOp::Eq => (self.0 == other.0).into_py(py),
+      CompareOp::Ne => (self.0 != other.0).into_py(py),
+      _ => py.NotImplemented(),
+    }
+  }
+
+  fn __hash__(&self) -> u64 {
+    self.0.hash.prefix_hash()
+  }
+
+  fn __repr__(&self) -> String {
+    format!("Digest('{}', {})", self.0.hash.to_hex(), self.0.size_bytes)
+  }
+}
+
+// -----------------------------------------------------------------------------
+// Snapshot
+// -----------------------------------------------------------------------------
+
+#[pyclass]
+struct PySnapshot(Snapshot);
+
+#[pymethods]
+impl PySnapshot {
+  #[new]
+  fn __new__() -> Self {
+    Self(Snapshot::empty())
+  }
+
+  #[classmethod]
+  fn _create_for_testing(
+    _cls: &PyType,
+    py_digest: PyDigest,
+    files: Vec<String>,
+    dirs: Vec<String>,
+  ) -> Self {
+    let snapshot = unsafe { Snapshot::create_for_testing_ffi(py_digest.0, files, dirs) };
+    Self(snapshot)
+  }
+
+  #[getter]
+  fn digest(&self) -> PyDigest {
+    PyDigest(self.0.digest)
+  }
+
+  #[getter]
+  fn files(&self, py: Python) -> Py<PyTuple> {
+    let files = self
+      .0
+      .path_stats
+      .iter()
+      .filter_map(|ps| match ps {
+        PathStat::File { path, .. } => path.to_str(),
+        _ => None,
+      })
+      .map(|ps| PyString::new(py, ps))
+      .collect::<Vec<_>>();
+    PyTuple::new(py, files).into()
+  }
+
+  #[getter]
+  fn dirs(&self, py: Python) -> Py<PyTuple> {
+    let dirs = self
+      .0
+      .path_stats
+      .iter()
+      .filter_map(|ps| match ps {
+        PathStat::Dir { path, .. } => path.to_str(),
+        _ => None,
+      })
+      .map(|ps| PyString::new(py, ps))
+      .collect::<Vec<_>>();
+    PyTuple::new(py, dirs).into()
+  }
+}
+
+#[pyproto]
+impl PyObjectProtocol for PySnapshot {
+  fn __richcmp__(&self, other: PyRef<PySnapshot>, op: CompareOp) -> Py<PyAny> {
+    let py = other.py();
+    match op {
+      CompareOp::Eq => (self.0.digest == other.0.digest).into_py(py),
+      CompareOp::Ne => (self.0.digest != other.0.digest).into_py(py),
+      _ => py.NotImplemented(),
+    }
+  }
+
+  fn __hash__(&self) -> u64 {
+    self.0.digest.hash.prefix_hash()
```

Comment:
Does this hash need the size mixed into the hash as well? While there should be only one size, as a protection against a bad size getting introduced we would probably want ('abc123', 5) and ('abc123', 10) to hash differently, correct?

Reply:
I'm copying the Rust-CPython code exactly. CC @stuhood if this is fine